boto/endpoints.py

```diff
@@ -0,0 +1,415 @@
+        if region_name is None:
+            if 'partitionEndpoint' in service_data:
+                region_name = service_data['partitionEndpoint']
+            else:
+                raise NoRegionError()
+        # Attempt to resolve the exact region for this partition.
+        if region_name in service_data['endpoints']:
+            return self._resolve(
+                partition, service_name, service_data, region_name)
+        # Check to see if the endpoint provided is valid for the partition.
+        if self._region_match(partition, region_name):
+            # Use the partition endpoint if set and not regionalized.
+            partition_endpoint = service_data.get('partitionEndpoint')
+            is_regionalized = service_data.get('isRegionalized', True)
+            if partition_endpoint and not is_regionalized:
+                LOG.debug('Using partition endpoint for %s, %s: %s',
+                          service_name, region_name, partition_endpoint)
+                return self._resolve(
+                    partition, service_name, service_data, partition_endpoint)
+            LOG.debug('Creating a regex based endpoint for %s, %s',
+                      service_name, region_name)
+            return self._resolve(
+                partition, service_name, service_data, region_name)
+
+    def _region_match(self, partition, region_name):
+        if region_name in partition['regions']:
+            return True
+        if 'regionRegex' in partition:
+            return re.compile(partition['regionRegex']).match(region_name)
+        return False
+
+    def _resolve(self, partition, service_name, service_data, endpoint_name):
+        result = service_data['endpoints'].get(endpoint_name, {})
+        result['partition'] = partition['partition']
+        result['endpointName'] = endpoint_name
+        # Merge in the service defaults then the partition defaults.
+        self._merge_keys(service_data.get('defaults', {}), result)
+        self._merge_keys(partition.get('defaults', {}), result)
+        hostname = result.get('hostname', DEFAULT_URI_TEMPLATE)
+        result['hostname'] = self._expand_template(
+            partition, result['hostname'], service_name, endpoint_name)
+        if 'sslCommonName' in result:
+            result['sslCommonName'] = self._expand_template(
+                partition, result['sslCommonName'], service_name,
+                endpoint_name)
+        result['dnsSuffix'] = partition['dnsSuffix']
+        return result
+
+    def _merge_keys(self, from_data, result):
+        for key in from_data:
+            if key not in result:
+                result[key] = from_data[key]
+
+    def _expand_template(self, partition, template, service_name,
+                         endpoint_name):
+        return template.format(
+            service=service_name, region=endpoint_name,
+            dnsSuffix=partition['dnsSuffix'])
+
+
+class BotoEndpointResolver(EndpointResolver):
```

Comment:
I went ahead and made `is_global_service` private. I had left it public to have tests for it since I added it before it was used. The particular edge cases are now covered by other tests, however.